Chartype fix

dr7ana opened the pull request:
@jagerman description:

- libc++-19 no longer supports `std::char_traits<T>` for `T=unsigned char` or `T=std::byte` (or, basically, anything other than the standard basic `char`, `char8/16/32_t`, `wchar_t`.  But since `std::basic_string<Char, Traits>` and `std::basic_string_view<Char, Traits>` default that second template argument to `class Traits = std::char_traits<Char>`.

- That means that our `bstring_view` and `ustring_view` are broken under llvm-19 (and so will break under macos once xcode upgrade to llvm-19.  God only knows what Apple will rename the version to when they do their usual llvm version mutilation deep into the internals of libc++, because Apple is a marketing company not a dev company).

Proposed solution:
- switch all our `ustring_view`s to `span<const unsigned char>`, our `bstring_view`s to `span<const std::byte>`, our `ustring` to `std::vector<unsigned char>`, and our `bstring`s to `std::vector<std::byte>`
- follow-up commit: do span<const T> for things we hard-code and want to leave read-only, and type-safe it as a const span (`cspan`). Span doesn't have comparators, they were added and then removed because of the shallow copy/deep compare asymmetry. So we type enforce constness for what the span already wraps, and then can define non-member comparators without having to pollute the std library well

dr7ana committed:
move _hex_u and _hex_b to span types

dr7ana committed:
better cspan typdefs, literals implemented, plus namespaced operators

dr7ana committed:
template specialization -> concept

dr7ana committed:
Test-cases implemented, spans are operational

dr7ana commented on oxenc/common.h:
```diff
@@ -34,22 +35,47 @@ concept tuple_like = requires { std::tuple_size<Tuple>::value; };
 
+template <class InputIterator1, class InputIterator2, class Cmp>
+constexpr auto lexi_three_way_compare(
+        InputIterator1 __first1,
```
drop dunders

jagerman commented on oxenc/span.h:
```diff
@@ -0,0 +1,83 @@
+}}  // namespace types::span
+
+template <typename T, typename U = T::value_type>
+concept const_span_like = basic_char<U> && std::same_as<const_span<const U>, T>;
```
This concept returns true for spans, but is a compile error (rather than false) for non-spans (or, at least, things without a `::value_type`):
```C++
#include <span>
#include <iostream>

template <typename Char>
concept basic_char = sizeof(Char) == 1 && !
std::same_as<Char, bool> && (std::integral<Char> || std::same_as<Char, std::byte>);

template <typename T, typename U = T::value_type>
concept const_span_like = basic_char<U> && std::same_as<std::span<const U>, T>;

template <typename T>
const char* f() {
    if constexpr (const_span_like<T>)
        return "span";
    else
        return "not span";
}

int main() {
    std::cout << f<std::span<char>>() << "\n";
    std::cout << f<int>() << "\n";
}
```
yields:
```
$ g++ -std=c++20 foo4.cpp 
foo4.cpp: In instantiation of ‘const char* f() [with T = int]’:
foo4.cpp:22:24:   required from here
   22 |     std::cout << f<int>() << "\n";
      |                  ~~~~~~^~
foo4.cpp:14:19: error: ‘int’ is not a class, struct, or union type
   14 |     if constexpr (const_span_like<T>)
      |                   ^~~~~~~~~~~~~~~~~~
```

dr7ana replied:
fixed, chatted off band

jagerman commented on oxenc/bt_serialize.h:
```diff
@@ -731,6 +751,20 @@ class bt_list_consumer {
         return {reinterpret_cast<const Char*>(result.data()), result.size()};
     }
 
+    template <typename Char = const char>
+    const_span<Char> consume_span() {
+        static_assert(std::is_const_v<Char>, "Span type must be const-qualifified!");
```
Using a static_assert here feels a bit less nice than using a `requires` (or concept) in the template to enforce the template type requirement.

But also, the `const` enforcement here (and on other similar functions) should be implied, not required-explicit, because the return value can never be anything other than a `span<const CharType>` (the consumer has read-only access to the buffer being parsed).

The "you must be explicit about const" isn't nice from an API perspective (i.e. for a user of oxenc) when there is no choice, because these consumer classes imply by their very nature that they can only ever return a span<const Char>.  "You must put the const keyword" doesn't add anything, and doesn't match things like `std::basic_string_view<char>` (or `consume_string_view<char>`) which give you back a string view with a `value_type = const char`, but doesn't require you (to put that `const` qualifier in the template argument because, just like this case, it's implied by the functionality.

If it worked like this:
```C++
auto sp = consumer.consume_span<std::byte>();
// sp is a span<const std::byte>
```
it seems like nothing would be lost in the API, but there's less typing and fewer checks to get there.

dr7ana replied:
Will change! Small note, `::value_type` never returns const-qualified, so the `std::basic_string_view<char>` has `::value_type = char`

jagerman commented on oxenc/bt_serialize.h:
```diff
@@ -912,6 +946,8 @@ class bt_list_consumer {
             throw bt_deserialize_invalid{"Dict finished without consuming the entire buffer"};
     }
+
+    using value_type = bt_value;
```
This won't work because `bt_value` (deliberately) isn't visible via this header outside of debug builds.  But also, a `value_type` typedef being here seems confusing because this isn't anything like a traditional container of homogeneous value types (where you'd usually find a value_type).

I'm also not seeing why this is needed/wanted.

dr7ana replied:
I added this so it would be workable with concepts that use `::value_type`. Also maybe I'm misunderstanding, but it still accesses that type in release builds?

jagerman commented on oxenc/span.h:
```diff
@@ -0,0 +1,83 @@
+inline namespace types { inline namespace span {
+    template <typename T, size_t N = std::dynamic_extent>
+    requires std::is_const_v<T> && std::equality_comparable<T>
+    using const_span = std::span<T, N>;
```
There's a hidden `equality_comparable` requirement on every use of `const_span`: the name should reflect that requirement (e.g. `comparable_const_span` or something).  With such a change, it then becomes apparently that this equality comparable requirement is being applied to every usage of const_span, even in places where equality comparable isn't required (and those should probably be slightly simpler `std::span<const C>`s instead).

But I'm wondering if this equality comparable check is actually needed at all: the only place I can see that needs equality_comparable on the spanned type are the two operators (which don't even use const_span to begin with).

dr7ana replied:
Yeah you're right, this is superfluous after our talk about genericizing the operators. changed!

jagerman commented on oxenc/bt_serialize.h:
```diff
@@ -731,6 +751,20 @@ class bt_list_consumer {
         return {reinterpret_cast<const Char*>(result.data()), result.size()};
     }
 
+    template <typename Char = const char>
```
This should be a `basic_char` instead of generic `typename` (and likewise for similar other calls), so that external code that does something invalid such as `c.consumer_span<int>()` is correctly diagnosed as a failure at the external invocation site, rather than a failure from deeper within oxenc.

dr7ana replied:
good point!

jagerman commented on oxenc/bt_serialize.h:
```diff
@@ -731,6 +751,20 @@ class bt_list_consumer {
     }
 
+    template <typename Char = const char>
+    const_span<Char> consume_span() {
```
It's not clear to me why the return type is `const_span<const Char>` and not `std::span<const Char>`.

Conceptually they are *almost* the same, except that `const_span` imposes a equality-comparable Char, but that doesn't seem necessary for this function: if I create some non-equality comparable custom char type that still satisfies `basic_char`, I should still be allowed to use it here.

dr7ana replied:
equality comparable dropped!

jagerman commented on oxenc/common.h:
```diff
@@ -14,17 +16,16 @@ concept string_view_compatible = std::convertible_to<T, std::string_view> ||
-concept basic_char = sizeof(Char) == 1 && !std::same_as<Char, bool> &&
-                     (std::integral<Char> || std::same_as<Char, std::byte>);
+concept basic_char = sizeof(Char) == 1 && !
+std::same_as<Char, bool> && (std::integral<Char> || std::same_as<Char, std::byte>);
```
This is some nasty reformatting by clang-format; we should bump format.sh's clang-format version to 16 instead of 15 so that it doesn't mess this up like this.  (There are other similar formatting abominations, such as the one in rlp_serialize.h).

jagerman commented on oxenc/common.h:
```diff
@@ -34,22 +35,47 @@ concept tuple_like = requires { std::tuple_size<Tuple>::value; };
-template <typename Char>
-inline constexpr bool is_string_like<std::basic_string_view<Char>> = sizeof(Char) == 1;
+template <typename T, typename U = T::value_type>
+concept string_like = basic_char<U> &&
```
See comment about `const_span_like` as to why the `::value_type` dependence in instantiation is a problem for a concept.

dr7ana replied:
Okay maybe this is a dumb question, but isn't the purpose of the template specialization (now a concept) to be (comment left unedited from dev):
```
// True if the type is a std::string, std::string_view, or some a basic_string<Char> for some
// single-byte type Char.
```
So wouldn't it make sense that if we want it constrain to string/view of some Char type, we could assume that there is a `::value_type`?

dr7ana replied:
Also, don't spots like [here](https://github.com/oxen-io/oxen-encoding/blob/dev/oxenc/bt_serialize.h#L298) assume a valid `::value_type`?

jagerman replied:
> So wouldn't it make sense that if we want it constrain to string/view of some Char type, we could assume that there is a ::value_type?

We currently use `is_string_like` in a couple `if constexpr`s in the code, which means dropping `is_string_like` in favour of just the concept without any backing template means the concept needs to be able to return true/false for any generic type, but currently it's true/compilation error.  You had to add `value_type` typedefs in a couple places to work around that compilation error, but those are workarounds for a limitation of the concept that didn't exist previously.

Having a `string_like` concept is perfectly fine, but trying to do it without partial specialization doesn't seem workable.  This is a simple and robust pattern:

```C++
template <typename T>
inline constexpr bool is_whatever = false;
template <typename T>
inline constexpr bool is_whatever<Whatever<T>> = true;

template <typename T>
concept whatever = is_whatever<T>;
```

because concepts do not allow, by design, template specialization: once your type requirements get complicated enough to warrant it, concept + specialization does the job perfectly well, keeping `concept` quite a bit simpler by not having any specialization logic in it.

There's a separate but related problem here, though, and it is that the negative path of the concept as written will instantiating the `std::string_view<T>/std::string_view<T>` just to evaluate the concept (which is likely a libc++ 19 issue).  But it's also forcing the compiler to "see" an instantiation that we may not care about or ever use, because if I have something like a `std::vector<int>` that passes through a `if constexpr (string_like<T>)` then it will compile (because there is a `::value_type`) but won't enter, but the still had the instantiate the `std::basic_string<int>` and `std::basic_string_view<int>`  classes along the way just to be able to evaluate it.

With template specialization that never happens: because `if constexpr (string_like<T>)` is simply false because `vector<int>` doesn't find any specialization and so goes straight to the `false` fallback.

There are other ways you might go about solving it, but I think it's pretty much always going to involve template specialization because what we're trying to match here is "any basic_string<T> or basic_string_view<T> that has a size-1 T", and that's basically a template matching problem.

jagerman replied:
> Also, don't spots like [here](https://github.com/oxen-io/oxen-encoding/blob/dev/oxenc/bt_serialize.h#L298) assume a valid `::value_type`?

Yeah, that looks wrong too; that one is definitely on me from embracing concepts perhaps a bit over-zealously.

There *is* a value way to check on a `::value_type` with a `requires (T) { ... }`, because a `requires` converts does-not-compile into false, and so some cases might be fixable with that, but the string_like I'm thinking is still going to need some specialization.

dr7ana replied:
okay, makes sense. thank you for the writeup. changed after chatting off-band as well

jagerman commented on oxenc/span.h:
```diff
@@ -0,0 +1,83 @@
+    template <typename T, size_t N, const_contiguous_range_t<T> R>
+    inline bool operator==(const std::span<T, N>& lhs, const R& rhs) {
+        return std::ranges::equal(lhs, rhs);
+    }
```
I think this should be enforcing std::equality_comparable on T instead of generic `typename`:
```suggestion
    template <std::equality_comparable T, size_t N, const_contiguous_range_t<T> R>
    inline bool operator==(const std::span<T, N>& lhs, const R& rhs) {
        return std::ranges::equal(lhs, rhs);
    }
```

jagerman commented on oxenc/span.h:
```diff
@@ -0,0 +1,83 @@
+        return std::ranges::equal(lhs, rhs);
+    }
+
+    template <typename T, size_t N, const_contiguous_range_t<T> R>
```
And this one should be enforcing threeway-comparability of `T`:
```suggestion
    template <std::three_way_comparable T, size_t N, const_contiguous_range_t<T> R>
```

jagerman commented on oxenc/common.h:
```diff
@@ -34,22 +35,47 @@ concept tuple_like = requires { std::tuple_size<Tuple>::value; };
 
 using namespace std::literals;
 
+template <class InputIterator1, class InputIterator2, class Cmp>
+constexpr auto lexi_three_way_compare(
+        InputIterator1 first1,
+        InputIterator1 last1,
+        InputIterator2 first2,
+        InputIterator2 last2,
+        Cmp comp) -> decltype(comp(*first1, *first2)) {
+#if defined(__clang__) && defined(_LIBCPP_VERSION) && (_LIBCPP_VERSION < 170000)
+    size_t len1 = last1 - first1;
+    size_t len2 = last2 - first2;
+    size_t minLen = std::min(len1, len2);
+    for (size_t i = 0; i < minLen; ++i, ++first1, ++first2) {
+        auto c = comp(*first1, *first2);
+        if (c != 0) {
+            return c;
+        }
+    }
+    return len1 <=> len2;
+#else
+    return std::lexicographical_compare_three_way(first1, last1, first2, last2, comp);
+#endif
+}
```
A preferable way to add a backport like this is to import the std function into your namespace directly when available, and then provide the alternative when not via something like this:
```C++
#if !defined(_LIBCPP_VERSION) || _LIBCPP_VERSION >= 170000
using std::lexicographical_compare_three_way;
#else
template <class InputIterator1, class InputIterator2, class Cmp>
constexpr auto lexicographical_compare_three_way(
        // ...
}
#endif
```
for two reasons:
- the std::-wrapping version doesn't have to make copies of the iterators.
- the "implement this missing std function" intention is a bit clearer when you directly use it rather than wrap around it.

jagerman replied:
Also: I dropped the `defined(__clang__)` check, because technically it's possible (albeit issue-prone) to use libc++ with a different compiler, and this is really a libc++ omission rather than a clang omission.

dr7ana replied:
Thank you! That is much better, changed

jagerman commented on oxenc/bt_producer.h:
```diff
@@ -308,6 +314,15 @@ class bt_list_producer {
 
+    /// Appends an element containing const_span data
+    template <const_span_like T>
+    void append(const T& data) {
```
There's no need to restrict this to const spans, but there *is* a need to restrict this to basic_char spans:
```suggestion
    template <basic_char T>
    void append(const std::span<T>& data) {
```

jagerman commented on oxenc/common.h:
```diff
@@ -14,17 +16,16 @@ concept string_view_compatible = std::convertible_to<T, std::string_view> ||
 
 template <typename Char>
-concept basic_char = sizeof(Char) == 1 && !std::same_as<Char, bool> &&
-                     (std::integral<Char> || std::same_as<Char, std::byte>);
```
I just spotted a bug in this concept (not added in this PR): this is not properly handling `const bool` and `const std::byte`.  I think it should be:
```C++
concept basic_char = sizeof(Char) == 1 && !std::same_as<std::remove_cv_t<Char>, bool> &&
                     (std::integral<Char> || std::same_as<std::remove_cv_t<Char>, std::byte>);
```

jagerman commented on oxenc/hex.h:
```diff
@@ -287,11 +287,14 @@ namespace detail {
         }
 
         Char decoded[N / 2 + 1];  // Includes a null byte so that view().data() is a valid c string
+        using size = std::integral_constant<size_t, N / 2>;
```
A `static inline constexpr size_t size = N / 2;` would be preferable to bringing integral_constant into things here.

Also, if that constant were moved up a line then it can be `Char decoded[size + 1];` (rather than having N/2 on two different lines).  I'm more or less indifferent about using a constant here, but if there is one, let's use it everywhere it applies.

jagerman commented on oxenc/hex.h:
```diff
@@ -287,11 +287,14 @@ namespace detail {
-            return {decoded, valid ? sizeof(decoded) - 1 : 0};
-        }
+
+        constexpr std::basic_string_view<Char> view() const { return {decoded, size::value}; }
```
Dropping the ternary on `valid` here (and in the span implementation) feels mildly incorrect because this is going to be a view (or span) into N/2 null characters.

But this is totally optional and I don't feel strongly about it (because there's also a static_assert on `.valid` in the the actual literal that calls this, and this is in the private `detail` namespace.)

dr7ana replied:
Yes my thoughts exactly, the static_assert made this kindof redundant

jagerman commented on oxenc/hex.h:
```diff
@@ -287,11 +287,14 @@ namespace detail {
-            return {decoded, valid ? sizeof(decoded) - 1 : 0};
-        }
+
+        constexpr std::basic_string_view<Char> view() const { return {decoded, size::value}; }
```
I think this `view()` just existing is going to be a problem with libc++ 19, because it'll still be instantiating a `basic_string_view<std::byte>` or whatever just because the function signature exists (whether or not we actually call it).  I think instead we can just drop the view() method entirely, and then have the `operator""_hex` just do a `return {Hex.data, Hex.size};` rather than calling `view()`.

jagerman replied:
(Not 100% sure on that; we should bump the `full_llvm(16)` build in drone to 19 and see)

jagerman commented on oxenc/hex.h:
```diff
@@ -7,7 +7,7 @@
 #include <string_view>
```
There's also oxenc/base64.h and base32z.h that have equivalent literals and need the same span fix as hex.

jagerman reviewed (changes requested):
On the whole this is solid work to get things changed over to span.

I do have various review comments, left as inline comments, mostly around the template types involved.

dr7ana committed:
review fixes

dr7ana committed:
formatting, clang-format vbump 15->16

dr7ana committed:
cleaner concept

dr7ana committed:
added clang/llvm-19 CI workflow

dr7ana committed:
dropped append_impl overloads that make libc++19 sad

dr7ana committed:
Made it to mordor (libc++-19)

jagerman committed:
Move append into append_signature_helper

Fixes a dangling reference to the local variable in the returned
span/string_view.

dr7ana committed:
format

dr7ana committed:
consume signature fix

jagerman commented on oxenc/bt_producer.h:
```diff
@@ -80,8 +80,7 @@ namespace detail {
         } else
-            static_assert(
-                    false,
+            throw std::invalid_argument(
```
Errors detectable at compile time should do so, rather than be runtime errors.  If the `false` is causing trouble (apparently it's standard-compliant only from C++23 onwards), perhaps `std::is_void_v<SignFunc>` will work.

jagerman commented on oxenc/common.h:
```diff
@@ -32,24 +30,73 @@ concept bt_input_dict_container =
+template <typename ForwardIt, typename ItValueType = std::iterator_traits<ForwardIt>::value_type>
+concept ordered_pair_iterator =
+        std::forward_iterator<ForwardIt> && std::tuple_size<ItValueType>::value == 2 &&
+        requires(ForwardIt it) {
```
Unused requires expression argument.

jagerman commented on oxenc/common.h:
```diff
@@ -32,24 +30,73 @@ concept bt_input_dict_container =
+        std::forward_iterator<ForwardIt> && std::tuple_size<ItValueType>::value == 2 &&
+        requires(ForwardIt it) {
+            typename ItValueType::first_type;
+            typename ItValueType::second_type;
```
This concept isn't working as intended, partly because of this requires expression.

Here's the concept invoked with various types that should, in theory, pass; the ones marked with a `// *` are failing the static_assert:
```C++
static_assert(!ordered_pair_iterator<std::unordered_map<int, int>::iterator>);           // *
static_assert(!ordered_pair_iterator<std::unordered_map<int, int>::const_iterator>);     // *
static_assert(ordered_pair_iterator<std::map<int, int>::iterator>);
static_assert(ordered_pair_iterator<std::map<int, int>::const_iterator>);
static_assert(ordered_pair_iterator<std::vector<std::pair<int, int>>::iterator>);
static_assert(ordered_pair_iterator<std::vector<std::pair<int, int>>::const_iterator>);
static_assert(ordered_pair_iterator<std::vector<std::tuple<int, int>>::iterator>);       // *
static_assert(ordered_pair_iterator<std::vector<std::tuple<int, int>>::const_iterator>); // *
static_assert(ordered_pair_iterator<std::vector<std::array<int, 2>>::iterator>);         // *
static_assert(ordered_pair_iterator<std::vector<std::array<int, 2>>::const_iterator>);   // *
```
The first one is a problem (i.e. we aren't properly failing on unordered_maps).  This is because `::first_type` is a `const Key`, but `unordered_map<const Key, ...>` isn't the same type as`unordered_map<Key, ...>` and so the `is_same<std::unordered_map<...>>` is false for case 1.  That's easily fixed with a `std::remove_const_t`.

The second one is failing because we also want to check for a const_iterator (because if what you have is a `const unordered_map<K, V>&` then your `begin()`/`end()` will be `std::unordered_map<K,V>::const_iterator` rather than `::iterator`.  So this is easily fixed by checking both possible iterator types.

(There's another case I wish we could solve of using a `std::unordered_map<K, V, ...>` where the `...` is non-default: most notably, using a custom hasher, but I don't see any way we can get to that starting from the iterator subtype).

Here's my stab at solving all of these (this makes all the static_asserts pass), split into two concepts because it substantially reduces duplication across the two iterator possibilities:
```c++
template <typename It, typename K, typename V>
concept is_unordered_map_iterator =
        std::same_as<It, typename std::unordered_map<K, V>::iterator> ||
        std::same_as<It, typename std::unordered_map<K, V>::const_iterator>;

template <typename ForwardIt, typename ItValueType = std::iter_value_t<ForwardIt>>
concept ordered_pair_iterator =
        std::forward_iterator<ForwardIt> && std::tuple_size_v<ItValueType> == 2 &&
        !is_unordered_map_iterator<
                ForwardIt,
                std::remove_const_t<std::tuple_element_t<0, ItValueType>>,
                std::tuple_element_t<1, ItValueType>>;
```

jagerman commented on oxenc/bt_producer.h:
```diff
@@ -299,19 +334,38 @@ class bt_list_producer {
+
+    /// Appends an element containing const_span data
+    template <basic_char T>
+    void append(const std::span<T>& data) {
```
Should this not be a `std::span<const T> data` instead, so that it is invokable from (for example) a const vector?

jagerman replied:
For that matter: wouldn't a single `std::span<const T> data` be generic enough to *also* handle the string_view case?

jagerman replied:
Oh, never mind, I was wrong: the deduction on `T` means this will only match literal spans, not convertible-to-spans.

jagerman commented on oxenc/bt_producer.h:
```diff
@@ -299,19 +334,38 @@ class bt_list_producer {
+
     /// Appends an element containing binary string data
-    template <string_view_compatible T>
+    template <char_span_convertible T>
```
I think this should be `const_span_convertible`?

jagerman replied:
That goes for various other cases in this file with the same pattern (i.e. read-only inputs).

jagerman commented on oxenc/bt_producer.h:
```diff
@@ -299,19 +334,38 @@ class bt_list_producer {
+            throw std::logic_error{"Cannot append to list when a sublist is active"};
+        append_impl(data.data(), data.size());
+        append_intermediate_ends();
+    }
```
This overload seems unnecessary if the immediately following template becomes `const_span_convertible` instead of `char_span_convertible`

jagerman commented on oxenc/bt_producer.h:
```diff
@@ -299,19 +334,38 @@ class bt_list_producer {
+        append_intermediate_ends();
+    }
+
+    /// Appends an element containing const_span data
```
Comment seems incorrect: this function would accept any of `span<T>`, `const span<T>`, `span<const T>`, `const span<const T>`, but only the latter two of those is a `const_span`.

dr7ana committed:
span overloads, added CI lint check

jagerman committed:
Backwards compatibility fixes

Fixes compatibility with current live libsession-util code:

- allow signing and verify functions to have non-char basic_string_view
  arguments.
- fix a case where append_impl was being called by an (unhandled)
  non-char string_view.

jagerman committed:
Version bump for char/span changes

jagerman committed:
Move common bt-specific code into bt_common.h

This allows bt_producer.h to no longer include all of bt_serialize.h
just for the concepts they share.
